### ISSUE TITLE
Version 3: Update interfaces and nothing else

### ISSUE THREAD

**Issue by pmjones** [auraphp/Aura.SqlQuery#108]
There have been several questions lately regarding the fact that the interfaces do not match the classes. The disconnect is because the classes have had features added, but adding the relevant methods to the interface would be a BC break.

Perhaps the time has come to update the interfaces, and in so doing move to version 3, with no other functionality changes.

Thoughts?

**Comment by cdekok** [comment]
If you plan to change it you might want to consider changing the split with SubSelectInterface and SelectInterface it's a bit annoying you cannot get any typehint as you can only typehint on one of interfaces please consider extending the interface for ease of use with an IDE.

**Comment by pavarnos** [comment]
Yes please!
Also fix col() which is yellow in phpStorm with 2 arguments.

I have ended up creating a small wrapper class that fixes some annoyances and adds extra utility functions. For example, adding the ability to include columns from the join table that get automatically prefixed by the table name, a joinLeft() / joinInner(), and extra methods to ease working with subselects (not shown below). 

```php
class WrappedSelect {
   /** @var Aura\SqlQuery\Common\SelectInterface */
   private $select;

    /**
     * Adds a JOIN table and columns to the query.
     * @param string       $joinType  The join type: inner, left, natural, etc.
     * @param string       $spec      The table specification; "real_table" or "real_table AS alias" 
     * @param string       $condition Join on this condition.
     * @param array        $columns   columns as passed to cols()
     * @return $this
     */
    public function join($joinType, $spec, $condition = null, $columns = null)
    {
        $this->select->join($joinType, $spec, $condition);
        $this->addColumns($columns, $spec);
        return $this;
    }

    /**
     * @param string $spec      The table specification; "foo" or "foo AS bar".
     * @param string $condition Join on this condition.
     * @param array  $columns   columns as passed to cols()
     * @return $this
     */
    public function joinLeft($spec, $condition = null, $columns = [])
    {
        return $this->join('left', $spec, $condition, $columns);
    }

    /**
     * Adds a FROM element to the query; quotes the table name automatically.
     * @param string $table The table specification; "foo" or "foo AS bar".
     * @param array  $columns
     * @return $this
     */
    public function from($table, $columns = [])
    {
        $this->select->from($table);
        $this->addColumns($columns, $table);
        return $this;
    }

    /**
     * @param array  $columns
     * @param string $tableName table name
     */
    private function addColumns($columns, $tableName)
    {
        if (empty($columns)) {
            return;
        }
        if (strpos($tableName, ' ') !== false) {
            // if the table spec is a subselect or function or anything funky then just add the raw columns
            $this->cols($columns);
            return;
        }
        $resolved = [];
        foreach ($columns as $col => $alias) {
            if (is_string($col)) {
                $col = $tableName . '.' . $col;
            } else if (!preg_match('|[.("]|', $alias)) {
                // no function call or other weirdness
                $alias = $tableName . '.' . $alias;
            }
            $resolved[$col] = $alias;
        }
        $this->cols($resolved);
    }

// ... etc
```

**Comment by pavarnos** [comment]
Can we also make several protected methods public and available via the interface? eg reset() and the build* functions, or provide getters for flags (for DISTINCT), groupBy, orderBy, having. 

I recently built a paginator class that builds a PageInformation data transfer object for a sliding window of pages. Its based on ZF1, ZF2, PagerFanta and jasongrimes/php-paginator. But none of the above support Aura, and seem quite complex for what is actually needed. With the above changes, i can make the paginator even cleaner / clearer and offer it to the Aura project if you want it. 

<img width="236" alt="paginator" src="https://cloud.githubusercontent.com/assets/589595/21082672/1344452a-c045-11e6-89cc-6c2b3a1c8de2.png">

**Comment by pavarnos** [comment]
can you create a v3 branch we can make pull requests against?

**Comment by harikt** [comment]
@pavarnos have created a 3.x branch.

**Comment by pavarnos** [comment]
a suggested "ship list" for v3?
- [x] #112 add missing public methods to select interface (and related eg ValuesInterface, Limit*Interface). Add missing public methods to UpdateInterface and WhereInterface (for #107)
- [x] merge SubSelectInterface into SelectInterface? (see https://github.com/auraphp/Aura.SqlQuery/issues/108#issuecomment-251047464: can not do this. Needed for https://github.com/auraphp/Aura.SqlQuery/blob/3.x/src/AbstractQuery.php#L403)
- [x] use variable argument lists (...$value) to make function arguments explicit (#111 for col(), #114 for having(), where(), orWhere() etc)
- [x] #100 add useReplace() sql command for MySQL (see #115)
- [x] fix as many Scrutinizer warnings as possible (some are pretty easy fixes) (#117)
- [x] add documentation for additional public methods (eg leftJoin() which I didn't realise existed until digging in: see my earlier comments on this issue)

I'm OK with doing all this: i have time this week. @harikt has been great at responding quickly.

**Comment by harikt** [comment]
@pavarnos I would suggest you to create separate PR's . Small PR's are easy to review and merge. Changes to interfaces in PR may take some time for review and getting merged.

I will try my best to respond / merge with my knowledge.

Thank you.

**Comment by harikt** [comment]
By the way the docs structure will change. I will make the changes soon.

**Comment by syrm** [comment]
Don't forget to add hasCols and getCols in interface ;)

**Comment by pavarnos** [comment]
@oxman see #112: just waiting on approval / merge. Its a big PR though...

**Comment by pmjones** [comment]
@pavarnos et al: I've just merged a bunch of 3.x-related stuff; let me know how much that moves the checklist forward.

**Comment by pavarnos** [comment]
thanks for all the merges. updated the checklist. i'm short of time to contribute much for the next wee while...

**Comment by pmjones** [comment]
@pavarnos No hurry on my part. Thanks to you and @harikt for all your work here!

**Comment by pmjones** [comment]
My apologies, folks -- my recent PRs to the 3.x branch have been over-and-above this plan. After the Quoter PR I will cease, and move on to documentation.

Further, I think the Subselect interface *can* be merged to SelectInterface, so long as we typehint that particular bit of logic to SelectInterface instead of Subselect.

Thoughts?

**Comment by pavarnos** [comment]
Sounds good

**Comment by harikt** [comment]
UPDATE : I have made the 3.x as default branch.

**Comment by pavarnos** [comment]
all the things in this issue have now been resolved? can it be closed?

**Comment by pmjones** [comment]
Looks like it.
